pkg/tbtc/node.go

```diff
@@ -1,17 +1,222 @@
+
+	// Chain rejects the result if it has less than 25% safety margin.
+	// If there are not enough signatures to preserve the margin, it does not
+	// make sense to submit the result.
+	signatureThreshold := config.HonestThreshold +
+		(config.GroupSize-config.HonestThreshold)/2
+	if len(signatures) < signatureThreshold {
+		return fmt.Errorf(
+			"could not submit result with [%v] signatures for signature threshold [%v]",
+			len(signatures),
+			signatureThreshold,
+		)
+	}
+
+	onSubmittedResultChan := make(chan uint64)
+
+	subscription := drs.chain.OnDKGResultSubmitted(
+		func(event *DKGResultSubmittedEvent) {
+			onSubmittedResultChan <- event.BlockNumber
+		},
+	)
+
+	returnWithError := func(err error) error {
+		subscription.Unsubscribe()
+		close(onSubmittedResultChan)
+		return err
+	}
+
+	groupPublicKeyBytes, err := result.GroupPublicKeyBytes()
+	if err != nil {
+		return returnWithError(
+			fmt.Errorf(
+				"could not extract public key bytes from the result: [%v]",
+				err,
+			),
+		)
+	}
+
+	alreadySubmitted, err := drs.chain.IsDKGResultSubmitted(groupPublicKeyBytes)
+	if err != nil {
+		return returnWithError(
+			fmt.Errorf(
+				"could not check if the result is already submitted: [%v]",
+				err,
+			),
+		)
+	}
+
+	// Someone who was ahead of us in the queue submitted the result. Giving up.
+	if alreadySubmitted {
+		return returnWithError(nil)
+	}
+
+	// Wait until the current member is eligible to submit the result.
+	eligibleToSubmitWaiter, err := drs.waitForSubmissionEligibility(
+		startBlockNumber,
+		memberIndex,
+	)
+	if err != nil {
+		return returnWithError(
+			fmt.Errorf("wait for eligibility failure: [%v]", err),
+		)
+	}
+
+	for {
+		select {
+		case blockNumber := <-eligibleToSubmitWaiter:
+			// Member becomes eligible to submit the result.
+			subscription.Unsubscribe()
+			close(onSubmittedResultChan)
+
+			logger.Infof(
+				"[member:%v] submitting DKG result with public key [0x%x] and "+
+					"[%v] supporting member signatures at block [%v]",
+				memberIndex,
+				result.GroupPublicKeyBytes,
+				len(signatures),
+				blockNumber,
+			)
+
+			return drs.chain.SubmitDKGResult(
+				memberIndex,
+				result,
+				signatures,
+			)
+		case blockNumber := <-onSubmittedResultChan:
+			logger.Infof(
+				"[member:%v] leaving; DKG result submitted by other member at block [%v]",
+				memberIndex,
+				blockNumber,
+			)
+			// A result has been submitted by other member. Leave without
+			// publishing the result.
+			return returnWithError(nil)
+		}
+	}
+}
+
+// waitForSubmissionEligibility waits until the current member is eligible to
+// submit a result to the blockchain. First member is eligible to submit straight
+// away, each following member is eligible after pre-defined block step.
+func (drs *dkgResultSubmitter) waitForSubmissionEligibility(
+	startBlockNumber uint64,
+	memberIndex group.MemberIndex,
+) (<-chan uint64, error) {
+	// T_init + (member_index - 1) * T_step
```

Comment:
Let's remove that comment. It doesn't give any value.

Reply:
Done in 9d52784867c24c8553a9f7c8a73427bf0cf4ecd4.